src/Builder/FrontendDialectHelper.cpp

```diff
@@ -11,232 +11,203 @@
 // Helper methods for handling input ONNX models.
 //
 //===----------------------------------------------------------------------===//
+
 #include "llvm/ADT/SmallVector.h"
-#include <llvm/Support/Endian.h>
-#include <llvm/Support/SwapByteOrder.h>
+#include "llvm/Support/Endian.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SwapByteOrder.h"
 
 #include "src/Builder/FrontendDialectHelper.hpp"
 
 namespace onnx_mlir {
 
+// Parses unsigned number.
+size_t parseOffsetOrLength(const std::string &value) {
+  char *end = nullptr;
+  size_t offsetOrLength = strtoull(value.c_str(), &end, 0);
+  assert(end != value.c_str() && "failed to parse offset or length");
+  return offsetOrLength;
+}
+
+// Reads external data from file location specified in tensor proto.
+// See https://github.com/onnx/onnx/blob/main/docs/ExternalData.md
+std::unique_ptr<llvm::MemoryBuffer> readExternalData(
+    const std::string &externalDataDir, const onnx::TensorProto &tp) {
+  std::string location;
+  uint64_t offset = 0;
+  uint64_t length = -1; // MemoryBuffer uses -1 to mean infinity
+  for (const onnx::StringStringEntryProto &entry : tp.external_data()) {
+    assert(entry.has_key() && "external_data entry must have key");
+    assert(entry.has_value() && "external_data entry must have value");
+    if (entry.key() == "location") {
+      location = entry.value();
+    } else if (entry.key() == "offset") {
+      offset = parseOffsetOrLength(entry.value());
+    } else if (entry.key() == "length") {
+      length = parseOffsetOrLength(entry.value());
+    }
+  }
+  assert(!location.empty() && "missing external data location");
+  llvm::SmallVector<char> path(externalDataDir.begin(), externalDataDir.end());
+  llvm::sys::path::append(path, location);
+  auto bufferOrError = llvm::MemoryBuffer::getFileSlice(
+      path, length, offset, /*IsVolatile=*/false);
```

Comment:
llvm will in some cases create a memory buffer that points to the file content by mmap and thereby read without memory copies, just a small heap allocation of the MemoryBuffer metadata